### ISSUE TITLE
A collect_all Future API is wanted

### ISSUE THREAD

**Issue by xush6528** [pytorch/pytorch#38937]
## 🚀 Feature

```py
future_list = some_user_funct()

collected_fut = rpc.collect_all(futureu_list)

collected_fut.wait()
```

`collected_fut.wait()` should internally wait for all futures in the list.

## Motivation

For sharded embeddings, what we do is

```py
futs = []
for shard in shards:
  fut = rpc.rpc_async(shard.owner(), helper_that_runs_shard_forward, input)
  futs.append(fut)

# In a later user function
rets = [fut.wait() for fut in futs]
torch.sum(rets)
```

A cleaner API would be

```py
futs = []
for shard in shards:
  fut = rpc.rpc_async(shard.owner(), helper_that_runs_shard_forward, input)
  futs.append(fut)

collected_fut = rpc.wait_all(futs)

def callback(collected_fut):
  return torch.sum(collected_fut.wait())

collected_fut._then(callback)


# In a later user function
collected_fut.wait()
```

## Pitch

Follow the [Folly Future collectAll API](https://github.com/facebook/folly/blob/master/folly/futures/Future.h),
add a `future = rpc.collect_all(future_list)` API.

Could use https://github.com/pytorch/pytorch/pull/37311 as a reference.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528 @jjlilley @osalpekar

**Comment by mrshenli** [comment]
@kiukchung mentioned this feature as well. 

As we are going to put `Future` in `torch.futures`, we can also add this helper function there.

**Comment by mrshenli** [comment]
closed by #39119
